src/util-flow-rate.c

```diff
@@ -0,0 +1,626 @@
+    /* Update start_ts in case of initiating the revisit of buffer */
+    if ((frs->dir[direction].last_idx == frs->dir[direction].size - 1) &&
+            (frs->dir[direction].last_idx != idx)) {
+        start_ts = p_ts;
+        if (idx != 0) {
+            /* Update the sum */
+            FlowRateClearSumTill(frs, 0, idx, direction);
+            /* Consider current packet a new start of the ring */
+            idx = 0;
+        }
+    }
+    /* If the packet has come in the last open interval of time */
+    if (p_ts.secs - start_ts.secs < frs->dir[direction].size) {
+        SCLogDebug("Packet in the ts size window");
+        if (idx > frs->dir[direction].last_idx + 1) {
+            /* Index is not the same as last or the next so, the ring must be flushed for the items
+             * in between and sum updated */
+            FlowRateClearSumTill(frs, frs->dir[direction].last_idx + 1, idx, direction);
+            frs->dir[direction].buf[idx] += pkt_len;
+            /* Update the total sum */
+            frs->dir[direction].sum += pkt_len;
+        } else if ((idx == frs->dir[direction].last_idx) ||
+                   (idx == frs->dir[direction].last_idx + 1)) {
+            /* Index matches the last updated index in the ring buffer or is the next index in the
+             * buffer */
+            /* Add to the existing open time interval */
+            frs->dir[direction].buf[idx] += pkt_len;
+            /* Update the total sum */
+            frs->dir[direction].sum += pkt_len;
+        } else {
+            SCLogDebug("Revisiting the ring after a full round");
+            /* Index is revisited after a full round of the buffer */
+            uint64_t prev_byte_count = frs->dir[direction].buf[idx];
+            /* Overwrite the buffer */
+            frs->dir[direction].buf[idx] = pkt_len;
+            DEBUG_VALIDATE_BUG_ON(frs->dir[direction].sum < prev_byte_count);
+            /* Sum should get rid of previous count on the same index */
+            frs->dir[direction].sum += pkt_len - prev_byte_count;
+            frs->dir[direction].start_ts = p_ts;
+        }
+        frs->dir[direction].last_idx = idx;
+    } else {
+        SCLogDebug("Packet outside the ring buffer size window");
+        /* Packet arrived after one or more rounds of the entire buffer */
+        /* Flush the entire buffer */
+        memset(frs->dir[direction].buf, 0, frs->dir[direction].size);
+        frs->dir[direction].last_idx = 0;
+        frs->dir[direction].start_ts = p_ts;
+        frs->dir[direction].buf[0] = pkt_len;
+        /* Overwrite the sum calculated so far */
+        frs->dir[direction].sum = pkt_len;
+    }
+    SCLogDebug("Byte sum for dir %d: %" PRIu64, direction, frs->dir[direction].sum);
+    /* In any case, update the last seen timestamp */
+    frs->dir[direction].last_ts = p_ts;
+}
+
+bool FlowRateIsExceeding(ThreadVars *tv, DecodeThreadVars *dtv, FlowRateStore *frs, int direction)
```

Comment:
feels a bit odd to update a counter here, seems more like something the caller should do